### ISSUE TITLE
BlitzCache -- Query Plan Hash in sys.dm_exec_query_stats vs Query Plan Hash in the query plan

### ISSUE THREAD

**Issue by BlitzErik** [BrentOzarULTD/SQL-Server-First-Responder-Kit#780]
**Do you want to request a *feature* or report a *bug*?**
Neither, this is reference for a possible Connect item

Connect Item filed: https://connect.microsoft.com/SQLServer/feedback/details/3129546

I've been trying to figure out recently why some plan costs were showing 0 in BlitzCache as part of #692 

I posted a question about them [here](http://dba.stackexchange.com/questions/166890/sql-server-query-plan-xml-queryplanhash-length) as well.

Digging in further on my system, I found some off behavior around zero-leading hashes. 

Here's some evidence:

```
SELECT deqs.query_hash, deqs.query_plan_hash, deqp.query_plan
FROM sys.dm_exec_query_stats AS deqs
CROSS APPLY sys.dm_exec_query_plan(deqs.plan_handle) AS deqp
WHERE deqs.query_plan_hash = 0x00ED538D26FCD990
OPTION(RECOMPILE)
```

When I run that query specifically for that Query Plan Hash, I get a few lines of results.

![image](https://cloud.githubusercontent.com/assets/11590415/24068105/0460586a-0b5d-11e7-8221-f0d9beeb7c92.png)

When I open the query plan, and display the XML, I see this:

![image](https://cloud.githubusercontent.com/assets/11590415/24068114/1ef894bc-0b5d-11e7-93a0-9a42219f3feb.png)

The Query Hash is fine, but the Query Plan Hash is missing the two leading zeroes after the 0x.

Similarly, when I run the plan cache query looking for 0x0346CE4D0DB617C1 (single leading 0), I find a plan:

![image](https://cloud.githubusercontent.com/assets/11590415/24068136/9b60a17a-0b5d-11e7-9dae-9a2f8b94f37b.png)

But in the plan XML, the leading 0 is missing again:

![image](https://cloud.githubusercontent.com/assets/11590415/24068142/a6d45bf0-0b5d-11e7-8a4c-b2f94afa1cfd.png)

Not sure where to go with this, or how to work around it. 

I could make some rules, of course, but I don't know if they'd always be correct.

If a Query Plan Hash is 18 chars long, leave it alone
If a Query Plan Hash is 17 chars long, add a single zero
If a Query Plan Hash is 16 chars long, add two zeroes

And on down the line, leaving out plans with a hash of 0x0000000000000000 which seem to occur for cursors only.

Boy is this annoying.

**Comment by peschkaj** [comment]
One (easy-ish) option is to always left pad with 18 zeroes and then take the right 18 characters and apply the conversion to `VARBINARY(MAX)` to that string. It's certainly faster than filing a bug and waiting for a fix to roll out.

**Comment by BlitzErik** [comment]
Yeah, I would only file the item so I'd have something to point to if things got weirder. Even if it got fixed, I'd have to leave the padding in because who the heck diddly darn knows when people would actually apply the patch.

**Comment by peschkaj** [comment]
"You need to upgrade your SQL Server installation to use this feature of sp_BlitzCache."

**Comment by BlitzErik** [comment]
Connected: https://connect.microsoft.com/SQLServer/feedback/details/3129546

**Comment by BlitzErik** [comment]
SUCCESS!

```
WITH XMLNAMESPACES('http://schemas.microsoft.com/sqlserver/2004/07/showplan' AS p)
SELECT   
		statement.value('sum(/p:StmtSimple/@StatementSubTreeCost)', 'float') QueryPlanCost,
		q.n.value('@QueryPlanHash', 'VARCHAR(128)') QPHW,
		'0X' + RIGHT('0000000000000000' + SUBSTRING(q.n.value('@QueryPlanHash', 'VARCHAR(18)'), 3, 18), 16)
, b.QueryPlanHash
FROM    #statements s
JOIN ##bou_BlitzCacheProcs b
ON s.QueryHash = b.QueryHash
CROSS APPLY s.statement.nodes('/p:StmtSimple') AS q(n)
WHERE q.n.value('@QueryPlanHash', 'VARCHAR(128)') IS NOT NULL
OPTION(RECOMPILE)
```

![image](https://cloud.githubusercontent.com/assets/11590415/24072722/9b35f81c-0bc1-11e7-8edf-9ea0ea3899a9.png)

**Comment by josemalv** [comment]
Column query_hash in sys.dm_exec_query_stats is Binary(8).
https://msdn.microsoft.com/en-us/library/ms189741.aspx

QueryHash and QueryPlanHash are string. 
http://schemas.microsoft.com/sqlserver/2004/07/showplan/sql2016sp1/showplanxml.xsd

<xsd:attribute name="QueryHash" type="xsd:string" use="optional"/>
<xsd:attribute name="QueryPlanHash" type="xsd:string" use="optional"/>

I think, you should solve datattype conversion.

**Comment by BlitzErik** [comment]
This will be closed in #781 and the Connect item was recently closed as fixed.
